src/main/java/org/jenkinsci/plugins/ParameterizedRemoteTrigger/RemoteBuildConfiguration.java

```diff
@@ -322,735 +333,949 @@ private String buildUrlQueryString(Collection<String> parameters) {
+            server.setAddress(this.remoteJenkinsUrl);
+
+        }
+
+        if (server == null) {
+            if(!isJobUrl) {
+                if(!isRemoteUrlSet && isRemoteNameEmpty)
+                    throw new AbortException("Configuration of the remote Jenkins host is missing.");
+                if(!isRemoteUrlSet && !isRemoteNameEmpty && globallyConfiguredServer == null)
+                    throw new AbortException(String.format(
+                                "Could get remote host with ID '%s' configured in Jenkins global configuration. Please check your global configuration.",
+                                this.remoteJenkinsName));
+            }
+            //Generic error message
+            throw new AbortException(String.format(
+                        "Could not identify remote host - neither via 'Remote Job Name or URL' (job:'%s'), globally configured"
+                        + " remote host (remoteJenkinsName:'%s') nor 'Override remote host URL' (remoteJenkinsUrl:'%s').",
+                        expandedJob, this.remoteJenkinsName, this.remoteJenkinsUrl));
+        }
+        return server;
+    }
+
     /**
      * Lookup up a Remote Jenkins Server based on display name
-     * 
+     *
      * @param displayName
      *            Name of the configuration you are looking for
-     * @return A RemoteSitez object
+     * @return A RemoteJenkinsServer object
      */
-    public RemoteJenkinsServer findRemoteHost(String displayName) {
-        RemoteJenkinsServer match = null;
-
+    private RemoteJenkinsServer findRemoteHost(String displayName) {
+        if(isEmpty(displayName)) return null;
+        RemoteJenkinsServer server = null;
         for (RemoteJenkinsServer host : this.getDescriptor().remoteSites) {
             // if we find a match, then stop looping
             if (displayName.equals(host.getDisplayName())) {
-                match = host;
+                server = host;
                 break;
             }
         }
+        return server;
+    }
 
-        return match;
+    protected static String removeTrailingSlashes(String string)
+    {
+        if(isEmpty(string)) return string;
+        string = string.trim();
+        while(string.endsWith("/")) string = string.substring(0, string.length() - 1);
+        return string;
+    }
+
+    protected static String removeQueryParameters(String string)
+    {
+        if(isEmpty(string)) return string;
+        string = string.trim();
+        int idx = string.indexOf("?");
+        if(idx > 0) string = string.substring(0, idx);
+        return string;
+    }
+
+    protected static String removeHashParameters(String string)
+    {
+        if(isEmpty(string)) return string;
+        string = string.trim();
+        int idx = string.indexOf("#");
+        if(idx > 0) string = string.substring(0, idx);
+        return string;
+    }
+
+    private String getRootUrlFromJobUrl(String jobUrl) throws MalformedURLException
```

Comment:
I assume you mean `@Nullable`, `@Nonnull`, `@CheckForNull`,...
Currently we use the ones from `javax.annotation`, based on JSR305. But since this JSR is dead I wonder which annotations are commonly used in Jenkins plugins.

Here some alternatives are listed: https://stackoverflow.com/questions/35892063/which-nonnull-java-annotation-to-use